litestar/_multipart.py

```diff
@@ -113,51 +68,55 @@ def parse_multipart_form(
-        headers: list[tuple[str, str]] = []
-
-        while line_end_index != -1:
-            line_end_index = form_part.find(b"\r\n", line_index)
-            form_line = form_part[line_index:line_end_index].decode("utf-8")
-
-            if not form_line:
-                break
-
-            line_index = line_end_index + 2
-            colon_index = form_line.index(":")
-            current_idx = colon_index + 2
-            form_header_field = form_line[:colon_index].lower()
-            form_header_value, form_parameters = parse_content_header(form_line[current_idx:])
-
-            if form_header_field == "content-disposition":
-                field_name = form_parameters.get("name")
-                file_name = form_parameters.get("filename")
-
-                if file_name is None and (filename_with_asterisk := form_parameters.get("filename*")):
-                    encoding, _, value = decode_rfc2231(filename_with_asterisk)
-                    file_name = unquote(value, encoding=encoding or content_charset)
-
-            elif form_header_field == "content-type":
-                content_type = form_header_value
-                content_charset = form_parameters.get("charset", "utf-8")
-            headers.append((form_header_field, form_header_value))
-
-        if field_name:
-            post_data = form_part[line_index:-4].lstrip(b"\r\n")
-            if file_name:
-                form_file = UploadFile(
-                    content_type=content_type, filename=file_name, file_data=post_data, headers=dict(headers)
-                )
-                fields[field_name].append(form_file)
-            elif post_data:
-                fields[field_name].append(post_data.decode(content_charset))
-            else:
-                fields[field_name].append(None)
+    chunk = await async_next(stream, b"")
+    if not chunk:
+        return fields
+
+    try:
+        with PushMultipartParser(boundary, max_segment_count=multipart_form_part_limit) as parser:
+            segment: MultipartSegment | None = None
+            data: UploadFile | bytes = b""
+            while not parser.closed:
+                for form_part in parser.parse(chunk):
+                    if isinstance(form_part, MultipartSegment):
+                        segment = form_part
+                        if segment.filename:
+                            data = UploadFile(
+                                content_type=segment.content_type or "text/plain",
+                                filename=segment.filename,
+                                headers=dict(segment.headerlist),
+                            )
+                    elif form_part:
+                        if isinstance(data, UploadFile):
+                            await data.write(form_part)
+                        else:
+                            data += form_part
+                    else:
+                        # end of part
+                        if segment is None:
```

Comment:
This cannot happen, `segment` will always have a value at this point. If not, then that would be a bug in PushMultipartParser. The contract is as follows: For each part of a valid multipart stream, the parser will emit exactly one `MultipartSegment` instance, followed by zero or more non-empty content chunks, followed by `None`. A segment cannot end before its header section is complete, that would be a `ParserError` exception.

Reply:
I know, this check mostly exists for the benefit of the type checker. One could argue that it should be an `assert` then, but since it's just an `is None` check I felt that there's not that much benefit to it (and we usually don't use the `assert` pattern elsewhere)